### DISCUSSION TITLE
Assertions on different properties of object

### DISCUSSION THREAD

**Discussion by mu88** [fluentassertions/fluentassertions#2263] [Q&A]
Hi 👋🏻

Please have a look at the following contrived example:
```c#
public class Person
{
    public required string Name { get; init; }
    
    public required DateTime Born { get; init; }
}

public class PersonService
{
    public Person Create(string name, DateTime born) => new() { Born = born, Name = name };
}

[Fact]
public void CreatePerson()
{
    // Arrange
    var testee = new PersonService();
    
    // Act
    var result = testee.Create("Dynamo", 12.April(1953));

    // Assert
    // does not compile: result.Name.Should().Be("Dynamo").And.Born.Should.BeAfter(10.April(1953));
    // does not compile: result.Should().Satisfy(
    //     person => person.Name.Should().Be("Dynamo"),
    //     person => person.Born.Should.BeAfter(10.April(1953)
    //     );
    result.Should().Match<Person>(person => person.Name == "Dynamo" && person.Born > 10.April(1953));
}
```

I'd like to make assertions on an object's properties of different types (here `string` and `DateTime`) and thereby make use of FluentAssertions' wonderful extensions, e. g. `.BeAfter()`.  
But all I came up with is the very last line of the snippet and I'm wondering whether there is a more elegant way so that I can keep using the type-specific assertions?

Thank you

**Comment by jnyrup** [comment]
For the contrived example, I would probably just go with
```cs
subject.Name.Should().Be("Dynamo");
subject.Born.Should().BeAfter(10.April(1953));
```

But I imagine the real scenario is more complex and the subject is a result of another assertion.
E.g. using `ContainSingle().Which`

Currently I've this kind of code several places 😞 
```cs
var person = persons.Should().ContainSingle().Which;
person.Name.Should().Be("Dynamo");
person.Born.Should().BeAfter(10.April(1953));
```

It would be nice to allow further chaining
```cs
persons.Should().ContainSingle()
    .Which.Should().Satisfy(p =>
    {
        p.Name.Should().Be("Dynamo");
        p.Born.Should().BeAfter(10.April(1953));
    });
```

**Reply by mu88** [reply]
But `person.Should().ContainSingle()` does not compile as it is not a collection

**Reply by jnyrup** [reply]
That's why I wrote `persons` in plural 😉

**Reply by mu88** [reply]
But how does that relate to my question? 😉

**Reply by jnyrup** [reply]
There's nothing wrong in writing
```cs
subject.Name.Should().Be("Dynamo");
subject.Born.Should().BeAfter(10.April(1953));
```

One doesn't need to chain _every_ assertion together.

The example with _persons_ being a `IEnumerable<Person>` is that not having a chainable API for doing multiple assertions on a item retrieved by `Which`, to my knowledge forces one to save the result of `Which` into a local variable and then write two separate assertions.

So what I'm trying to, is making a case where the "workaround" isn't pretty or even throw unexpected exceptions as shown in the example below.

```cs
using(new AssertionScope())
{
    var person = persons.Should().ContainSingle().Which;
    person.Name.Should().Be("Dynamo"); // <-- this would throw an NRE if ContainSingle failed
    person.Born.Should().BeAfter(10.April(1953));
}
```

**Reply by mu88** [reply]
Of course there's nothing wrong with writing it like that:
```c#
subject.Name.Should().Be("Dynamo");
subject.Born.Should().BeAfter(10.April(1953));
```

But think about a deeper object graph, e. g. `Person.Address.City.Major.Person` (being a `Person` as well). Now you can do the following:
```c#
// duplicate
Person.Address.City.Major.Person.Name.Should().Be("Dynamo");
Person.Address.City.Major.Person.Born.Should().BeAfter(10.April(1953));

// introduce temporary object
var person = Person.Address.City.Major.Person;
person.Name.Should().Be("Dynamo");
person.Born.Should().BeAfter(10.April(1953));
```

I'd like it if I could write:
```c#
Person.Address.City.Major.Person.Should().Satisfy(
    person => person.Name.Should().Be("Dynamo"),
    person => person.Born.Should().BeAfter(10.April(1953)
);
```

**Comment by jnyrup** [comment]
We have a similar API `AllSatisfy` for collections, we could probably mimic it just for non-collection subjects.

See it's usage 
https://github.com/fluentassertions/fluentassertions/blob/329060d060e1bb9c86171115559c5d9ac333d29d/Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.AllSatisfy.cs#L79-L117

**Comment by dennisdoomen** [comment]
Or use something like this:

```cs
  result.Should().BeEquivalentTo(new
  {
      Name = "Dynamo",
  }, options => options
      .Using<DateTime>(a => a.Subject.Should().BeAfter(10.April(1953)))
      .WhenTypeIs<DateTime>());
```
